src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs

```diff
@@ -36,21 +33,68 @@ public static class TopologicalSortUtility
             return sorted;
         }
 
-        private static int GetParentCount(List<PackageDependencyInfo> packages, string id)
+        private static void Sort(PackageInfo[] packages, int start)
+        {
+            // Update parent counts
+            for (var i = start; i < packages.Length; i++)
+            {
+                var package = packages[i].Package;
+                // Mutating stuct in-place, need to index rather than use enumerator
+                packages[i].ParentCount = GetParentCount(packages, package.Id, start);
+            }
+
+            Array.Sort(packages, start, packages.Length - start, _comparer);
+        }
+
+        private static int GetParentCount(PackageInfo[] packages, string id, int start)
         {
-            int count = 0;
+            var parentCount = 0;
 
-            foreach (var package in packages)
+            for (var i = start; i < packages.Length; i++)
             {
-                if (package.Dependencies != null
-                    && package.Dependencies.Any(dependency =>
-                        string.Equals(id, dependency.Id, StringComparison.OrdinalIgnoreCase)))
+                var deps = packages[i].Package.Dependencies;
+                var dependencies = deps as PackageDependency[] ?? deps.ToArray();
```

Comment:
I understand that this is casting if it is already an array, or making one. But does casting it here cause foreach to treat it differently than if it is an array behind the IEnumerable interface? How does this work?